### PR TITLE
{devel}[foss/2018b] PyTorch v1.0.1 w/ Python 3.6.6

### PR DESCRIPTION
**Pull request by branfosj**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in this PR)
bear-pg0212u15b.bear.cluster - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 2.7.5
See https://gist.github.com/0318ecbb443492446cb13e314f8ed2bd for a full test report.

**Comment by branfosj** [comment]
I've made the requested updates. However, it seems to have broken the PyTorch install...

File: easybuild/easyconfigs/h/hypothesis/hypothesis-4.5.0-foss-2018b-Python-3.6.6.eb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ by a source of examples, and then generates simple and comprehensible examples that make your tests fail. This lets
+ you find more bugs in your code with less work."""
+
+toolchain = {'name': 'foss', 'version': '2018b'}
+
+dependencies = [
+    ('Python', '3.6.6'),
+]
+
+use_pip = True
```
Comment (vanzod) [inline easybuild/easyconfigs/h/hypothesis/hypothesis-4.5.0-foss-2018b-Python-3.6.6.eb]: Missing `download_dep_fail = True`

File: easybuild/easyconfigs/h/hypothesis/hypothesis-4.5.0-fosscuda-2018b-Python-3.6.6.eb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ by a source of examples, and then generates simple and comprehensible examples that make your tests fail. This lets
+ you find more bugs in your code with less work."""
+
+toolchain = {'name': 'fosscuda', 'version': '2018b'}
+
+dependencies = [
+    ('Python', '3.6.6'),
+]
+
+use_pip = True
```
Comment (vanzod) [inline easybuild/easyconfigs/h/hypothesis/hypothesis-4.5.0-fosscuda-2018b-Python-3.6.6.eb]: Missing `download_dep_fail = True`

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-foss-2018b-Python-3.6.6.eb [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+homepage = 'http://pytorch.org/'
+description = """Tensors and Dynamic neural networks in Python with strong GPU acceleration.
+PyTorch is a deep learning framework that puts Python first."""
+
+toolchain = {'name': 'foss', 'version': '2018b'}
+
+pytorchdir = 'pytorch-%(version)s'
+extract_cmd_pattern = 'tar -C %s/%s --strip-components=1 -xf %%s'
+mklml_extract_cmd = "mkdir -p %s/%s && tar -C %s/%s --strip-components=1 -xzf %%s"
+
```
Comment (vanzod) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-foss-2018b-Python-3.6.6.eb]: Missing `use_pip = True` and `download_dep_fail = True`

File: easybuild/easyconfigs/p/Python/Python-3.6.6-fosscuda-2018b.eb [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    # it's nice to have an up to date OpenSSL for security reasons;
+    # ('OpenSSL', '1.1.0h'),
+]
+
+osdependencies = [('openssl-devel', 'libssl-dev', 'libopenssl-devel')]
+
+# Try to make sure stack limit is set to unlimited, to avoid compilation errors for hashlib with Intel compilers
+# cfr. https://github.com/easybuilders/easybuild-easyconfigs/issues/6484
+ulimit_unlimited = True
+exts_download_dep_fail = True
```
Comment (vanzod) [inline easybuild/easyconfigs/p/Python/Python-3.6.6-fosscuda-2018b.eb]: Missing `use_pip = True`

**Comment by branfosj** [comment]
I see the same as @vanzod does in https://github.com/easybuilders/easybuild-easyconfigs/pull/7717#issuecomment-467237727 so I've taken `use_pip = True` out of the PyTorch easyconfigs.

**Comment by akesandgren** [comment]
There is already a Ninja-1.8.2-fosscuda-2018b in PR https://github.com/easybuilders/easybuild-easyconfigs/pull/7654 although it uses Python 2.7.15 for its builddependency.
We could use that since it also include Meson.

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-fosscuda-2018b-Python-3.6.6.eb [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+]
+
+builddependencies = [('CMake', '3.12.1')]
+
+dependencies = [
+    ('Python', '3.6.6'),
+    ('PyYAML', '3.13', versionsuffix),
+    ('hypothesis', '4.5.0', versionsuffix),
+    ('cuDNN', '7.1.4.18'),
+    ('Ninja', '1.8.2'),
```
Comment (akesandgren) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-fosscuda-2018b-Python-3.6.6.eb]: Add "magma" to the dependencies? See https://github.com/easybuilders/easybuild-easyconfigs/pull/7717 for the fosscuda version there.

**Comment by akesandgren** [comment]
Btw, the patches from https://github.com/easybuilders/easybuild-easyconfigs/pull/7717 are they of any interest in this version?

**Comment by branfosj** [comment]
> 
> 
> Btw, the patches from #7717 are they of any interest in this version?

- `PyTorch-0.3.1-cmake.patch` - the file does not exist anymore 
- `PyTorch-0.3.1_skip_MPI_tests.patch` - will not apply and is not in 0.4.1 PyTorch version
- `PyTorch-0.3.1_fix_findAVX.patch` - I believe this should still be there, but it needs updating for the new version

patches from other versions:
- `PyTorch-0.3.1-fix-for-intel.patch` - this should be there if we want consistency between toolchains
- `PyTorch-0.3.1_intel_skip_autgrad_test.patch` - this test works for me and this is not in the 0.4.1 PyTorch version

End result is that I'll add the following - updating as needed:
- `PyTorch-0.3.1_fix_findAVX.patch`
- `PyTorch-0.3.1-fix-for-intel.patch`

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-fosscuda-2018b-Python-3.6.6.eb (hunk 5 above)
Comment (branfosj) [reply easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-fosscuda-2018b-Python-3.6.6.eb]: added

**Comment by branfosj** [comment]
Thanks for the info. I've not yet built the latest files here, as I'm waiting on availability of a system with a GPU to be able to test. I expect this means we should drop down to using magma 2.4.0 instead.

**Comment by dithwick** [comment]
No worries, I thought I'd mention it before you end up going through the same pain I did. I guess the alternative is to specify the sm_35, sm_50 etc. as GPU targets, looking at the Makefile it should accept those at options.

**Comment by branfosj** [comment]
> Merge branch 'develop' of github.com:easybuilders/easybuild-easyconfi…

I was about to fix the test issue, but I see that that'll come from #7844 

I have to get magma building, then I'll produce a build test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
nic158 - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/bfc9f58577f9e4dfde5a49c5617e5380 for a full test report.

**Comment by branfosj** [comment]
@boegel - I'll upload a new patch for magma shortly. I forgot to change the `?=` to `:=` in the Makefile patch.

**Comment by boegel** [comment]
@branfosj :+1:

**Comment by boegel** [comment]
@branfosj Maybe it's worth pulling this apart, and getting the `PyTorch` easyconfig using `foss/2018b` merged first (testing that is failing too for me btw, see upcoming test report)

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
node2402.golett.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/42f87453eadb2502ae0d98ef9469e0eb for a full test report.

**Comment by boegel** [comment]
@branfosj The `use_pip` complaints is being dealt with in #7844, should get merged soon into `develop`, which will fix the problem in the tests here on the next commit/test trigger.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
nic158 - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/64d7e9c9960f52aca404341a0eb40b0e for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
bear-pg0212u25a.bear.cluster - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 2.7.5
See https://gist.github.com/14aed09c7e30fa9c4e1c1525bd29fab0 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
bb2c1c10.bb2.cluster - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2660 0 @ 2.20GHz, Python 2.7.5
See https://gist.github.com/0593f0995bb0cc33935fdea8946ea180 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
nic158 - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/6f972303abfe2a99f4c379c02936a9a6 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bber0501u03b.bb2.cluster - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/35cb574328c53eaaae85c8b05429ff40 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in this PR)
node3102.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/de592aaea3486f28793bcce2dd28c2c1 for a full test report.

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-foss-2018b-Python-3.6.6.eb [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    ('PyYAML', '3.13', versionsuffix),
+    ('hypothesis', '4.5.0', versionsuffix),
+    ('Ninja', '1.8.2'),
+]
+
+download_dep_fail = True
+
+# This environment is needed to prevent setup.py from invoking git to determine version.
+prebuildopts = 'PYTORCH_BUILD_VERSION=%(version)s PYTORCH_BUILD_NUMBER=1 VERBOSE=1 '
+prebuildopts += ' CFLAGS="-ldl"'
```
Comment (wpoely86) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-foss-2018b-Python-3.6.6.eb]: change this to:

```python
prebuildopts += ' LDFLAGS="$LDFLAGS -ldl"'
preinstallopts = 'VERBOSE=1 LDFLAGS="$LDFLAGS -ldl"'

```

Comment (branfosj) [reply easybuild/easyconfigs/p/PyTorch/PyTorch-1.0.1-foss-2018b-Python-3.6.6.eb]: Thanks. Done.

**Comment by wpoely86** [comment]
It builds fine but I have one test that always fails:
```
======================================================================
FAIL: test_affine_3d_rotateRandom (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dev/shm/2034712/build/PyTorch/1.0.1/fosscuda-2018b-Python-3.6.6/pytorch-1.0.1/test/common_utils.py", line 117, in wrapper
    fn(*args, **kwargs)
  File "test_nn.py", line 6036, in test_affine_3d_rotateRandom
    assert np.allclose(affine_tensor[0, i, r, c], grid_out[:3], atol=1e-5)
AssertionError
```

**Comment by wpoely86** [comment]
Test report by @wpoely86
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
nic139 - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/c892be956015c998b96c49f33e8bf022 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
bear-pg0212u15b.bear.cluster - Linux centos linux 7.5.1804, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 2.7.5
See https://gist.github.com/3b19e251086cf844f8fd8ba6f92e4faa for a full test report.

**Comment by wpoely86** [comment]
OK, it worked this time. Not sure why. @boegel can you also retest it?

**Comment by akesandgren** [comment]
Test report by @akesandgren
**SUCCESS**
Build succeeded for 3 out of 3 (2 easyconfigs in this PR)
b-an03.hpc2n.umu.se - Linux ubuntu 16.04, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 2.7.12
See https://gist.github.com/61bf0160c6e1b97aca0e9f556c5b70e5 for a full test report.

**Comment by akesandgren** [comment]
@boegel @vanzod looks good to me...

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 3 (2 easyconfigs in this PR)
node2473.golett.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/521714f94ebeb989b153b7a02be81950 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
node2455.golett.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/caac65b0dfa479a19f033e0c885c09fe for a full test report.

**Comment by boegel** [comment]
Going in, thanks @branfosj!

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
node3105.skitty.os - Linux centos linux 7.6.1810, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 2.7.5
See https://gist.github.com/a040cd4d58c10cddeb8e6035c4ef33b9 for a full test report.

**Pull request merged**